---
title: "SNiPDx Integration"
subtitle: " `r params$currTrial`  `r params$currPatient` Samples"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
    code_download: true
    collapsed: yes
    highlight: tango
    number_sections: no
    theme: yeti
    toc: yes
    toc_float: yes
fontsize: 25pt
editor_options:
  chunk_output_type: console
params:
  inExtensions: inExtensions
  inInventory: inInventory
  inSamplesFolder: inSamplesFolder
  inVendorData: inVendorData
  currTrial: currTrial
  panalAnnotations: panalAnnotations
  currPatient: currPatient
  processingDir: processingDir
---
<style>
    body {
    	font-size: 16px;
        text-align: justify;
    }

    .main-container {
        max-width: 80%;
        margin-left: auto;
        margin-right: auto;
    }

    code {
    	font-size: 75%;
    }

    code.r {
    	font-size: 80%;
    }

    div.main-container {
        max-width: 80%;
    }
</style>

```{r, inputVar, echo = FALSE, eval=FALSE}
# ==== specify input variables here ====

#interactive session
params <- list()

# ### Clinical example
# params$currPatient <- "1004-0008" #test individual patient
# 
# params$currTrial <- "RP6306-02"
# 
# params$inExtensions <- here::here("/ClinBio/SP-ClinicalBioinformatics/ntran/analysis/SNIPDx/nf-snipdx-dev/nf-snipdx/src/utils/snipdx_reports.R")
# 
# params$inInventory <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/RP6306-02/vendor-data/INVENTORY/rp6306-02_genosity_inv_2022-10-10.csv")
# 
# params$inSamplesFolder <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/RP6306-02/vendor-data/dominik/RP6306-02/")
# 
# params$inVendorData <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/RP6306-02/vendor-data/ARCHER/")
# 
# params$panalAnnotations <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/git-repos/reparecompbio/clinicalbioinfoCLIs/snipdx/ref/panel_genes_annotated_V2.tsv")
#   

### Nonclinical example 
params$currPatient <- "MDACC_LMS_RNASEH2" #test individual project

params$currTrial <- "Non-Clinical"

params$inExtensions <- here::here("/ClinBio/SP-ClinicalBioinformatics/ntran/analysis/SNIPDx/nf-snipdx-dev/nf-snipdx/src/utils/snipdx_reports.R")

params$inInventory <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/Non-Clinical/vendor-data/gcp/INVENTORY/rp_nonclin_genosity_inv_annotated_2022-12-06.csv")

params$inSamplesFolder <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/Non-Clinical/vendor-data/gcp/ARCHER/")

params$inVendorData <- here::here("/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/Non-Clinical/vendor-data/gcp/ARCHER/")

params$panalAnnotations <- here::here("/ClinBio/SP-ClinicalBioinformatics/ntran/analysis/SNIPDx/nf-snipdx-dev/nf-snipdx/refs/panel_genes_annotated-v2.tsv")

params$processingDir <- "/ClinBio/SP-ClinicalBioinformatics/shared/SNiPDx/Non-Clinical/processing/"
  
```

```{r,setup, include=FALSE}

#global knit options
knitr::opts_chunk$set(warning=FALSE, message=FALSE, echo = FALSE)
#knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file()) 


library(tidyverse)
library(rmarkdown)
library(readxl)
library(openxlsx)
library(reshape2)
library(DT)
library(ggrepel)
library(here)
library(cowplot)
library(purrr)
library(data.table)
library(binom)
library(GenomicRanges)
library(magrittr)
library(dplyr)

# devtools::load_all(params$inExtensions)
source(params$inExtensions)

# if(params$currTrial != "Non-Clinical"){
#   folder_name <- paste0("./results/", Sys.Date(), "_", params$currTrial,
#                         "_", "SNiPDx_by_Patients")
# }else{
#   folder_name <- paste0("./results/", Sys.Date(), "_", params$currTrial,
#                         "_", "SNiPDx_by_Projects")
# }
# 
# script_name <- paste(Sys.Date(),params$currTrial, "SNiPDx", params$currPatient,
#                      sep = "_")
# 
# allScriptName <- paste(Sys.Date(),params$currTrial, 
#                                                "SNiPDx", "Complete",
#                                                sep = "_")
# 
# dataPath <- paste0("./data/", allScriptName, "_Inventory_Processed_Data.rds")

```

```{r, readIn,results='hide'}
### read data 
Inventory <- read_csv(params$inInventory)
inStats <- list.files(path = params$inVendorData,".Stats_Summary",full.names = T,recursive = T)
inVariant <- list.files(path = params$inVendorData,".variant_summary",full.names = T,recursive = T)
inSegment <- list.files(path = params$inSamplesFolder,"*.segments.csv", full.names = T,recursive = T)


### subset for SNIPDx samples only (remove WGS test code c("AA031a", "AA032a") from inventory)
Inventory <- Inventory %>% filter(`Test Code` %in% c("AA091a", "AA099a"))

### Read in Stats
statsList <- lapply(inStats, read_tsv)

sharedCols <-  Reduce(intersect,lapply(statsList, colnames))

statsSharedList <- lapply(statsList, select, sharedCols)

Stats <- do.call(rbind, statsSharedList)

### Read in Variant
# specify that gnomAD.Alt.Freq.All and `gnomAD Max Alt Freq` should always be of class double
# some cases where these columns are read in as a character due to sci notations
Variant <-
  inVariant %>% 
  map_df(~read_tsv(.,col_types = cols(gnomAD.Alt.Freq.All = col_double(),
                                      `gnomAD Max Alt Freq`= col_double()))) %>% 
# revalue MRE11A to MRE11 to match gene panel annotations 
mutate(Gene = case_when(Gene %in%  "MRE11" ~"MRE11A",
                                     TRUE ~ as.character(Gene)))

### Read in sample level Segments files per sample (older samples may not have segments file)
# match sample ID with segments file 

# remove segment file paths from facets_with_purecn_diplogr folder
inSegment <- inSegment[grep("facets_with_purecn_diplogr", inSegment, invert = T)]

segmentLists <- list()
# iterate thru segment files per sample
for(currSegment in inSegment){
  # extract sample seq id
  currSample <- str_match(currSegment, "facets/\\s*(.*?)\\s*.segments.csv")[,2]
  
  # remove the parent dir ("chromosome_plots/)
  currSample <- gsub("chromosome_plots/", "", currSample)
  
  # read in and append to sample sequencing name column, query sample id to inventory downstream
  seg_df <- read_csv(currSegment) %>%
    mutate(`Sample Sequencing Name` = currSample)
  
  #### DERIVE PIER SCORE at sample level (?) ####
   median.logr <- median(seg_df$mafR)
    std.logr <- sd(seg_df$mafR)
    seg_df$isLohPier <- seg_df$mafR > median.logr + std.logr
    
    seg_df$segments.loh.pier <- as.integer(seg_df$isLohPier)

  segmentLists[[currSample]] <- seg_df
}

OrigSegments <- rbindlist(segmentLists)
```

```{r, facets, results='hide'}

gene_table <- read.table(params$panalAnnotations, 
                         sep='\t', 
                         header = TRUE)

gene_table <- gene_table[!is.na(gene_table$ensembl_gene_id) &
                           !duplicated(gene_table$ensembl_gene_id),]

gene_table <- gene_table[order(gene_table$chromosome_name, gene_table$transcript_start),]
#gene_table$hgnc_symbol[gene_table$hgnc_symbol=='MRE11A'] <- 'MRE11'
gene_table$hgnc_symbol_chr <- paste0(gene_table$hgnc_symbol, ' chr',gene_table$chromosome_name)

gene.gr <- GRanges(seqnames=Rle(paste0('chr',gene_table$chromosome_name)),
                   ranges=IRanges(gene_table$transcript_start, gene_table$transcript_end),
                   strand=rep(c("*"), nrow(gene_table)))

facets.files <- list.files(path = params$inSamplesFolder,
                           pattern = "*.cnv.table.txt",
                           all.files = FALSE,
                           full.names = F,
                           recursive = TRUE,
                           ignore.case = FALSE, 
                           include.dirs = FALSE, 
                           no.. = FALSE) # 524 obs

# extract cnv tables from REPARE/ folder only (omit Genosity generated facets)
facets.files <- facets.files[grep("FACETS/REPARE/facets/", facets.files)]

call.list <- list()
# loop over all files
for (ti in 1:length(facets.files)) {	
  bn <- basename(facets.files[ti])
  full_sample_name <- gsub('.cnv.table.txt|.dominik.txt', '', bn)
  print(full_sample_name)
  split_vec <- strsplit(bn, '_')[[1]]
  ger_id <- split_vec[grepl('GER', split_vec)]
  print(ger_id)
  if (length(ger_id)==0) {
    ger_id=NA
  }
  
  
  # prepare the data frame that will hold the results for a single case
  sample.df <- data.frame(full_sample_id=full_sample_name,
                          ger.id = ger_id,
                          Gene=gene_table$hgnc_symbol,
                          transcript_start=gene_table$transcript_start,
                          transcript_end=gene_table$transcript_end,
                          chr=gene_table$chromosome_name,
                          facets.tcn.em.median=NA, facets.tcn.em.min=NA, facets.tcn.em.max=NA,
                          facets.lcn.em.median=NA,facets.lcn.em.min=NA,
                          facets.ploidy=NA, facets.purity=NA
  )
  
  # read the facets output
  # Facets V2
  # facets.calls <- try(read.table(inFacets[ti], header=TRUE, sep='\t'))
  
  facets.calls <- withr::with_dir(params$inSamplesFolder,{
  try(read.table(facets.files[ti], header=TRUE, sep='\t'))
  })

  
  
  
  if ((class(facets.calls) == "try-error") ) {
    print('Facets error') 
  } else {
    sample.df$facets.ploidy <- facets.calls$ploidy[1]
    sample.df$facets.purity <- facets.calls$purity[1]
    
    # fill in missing minor copy number, according to Pier's trick
    facets.calls$segmentSize <- facets.calls$end - facets.calls$start
    median.logr <- median(facets.calls$mafR)
    std.logr <- sd(facets.calls$mafR)
    facets.calls$isLohPier <- facets.calls$mafR > median.logr + std.logr
    
    facets.gr <-  GRanges(paste0('chr',seqnames=Rle(facets.calls$chr)),
                          ranges=IRanges(as.integer(facets.calls$loc.start),
                                         as.integer(facets.calls$loc.end)),
                          strand=rep(c("*"), nrow(facets.calls)))
    # perform the overlap between genes and facets segments
    # for each gene, there may be multiple overlapping segments
    r <- as.data.frame(findOverlaps(gene.gr, facets.gr)) 
    r$facets.tcn.em <- NA
    r$facets.tcn.em <- facets.calls$tcn.em[r$subjectHits]
    r$facets.lcn.em <- NA
    r$facets.lcn.em <- facets.calls$lcn.em[r$subjectHits]
    r$isLohPier <- NA
    r$isLohPier <- as.integer(facets.calls$isLohPier[r$subjectHits])
    
    # if there are any overlaps
    if (nrow(r)>0) {
      
      # total copy number: median, min and max
      facets.tcn.em.median = aggregate(r$facets.tcn.em,
                                       by = list(r$queryHits),
                                       FUN = median, na.rm=TRUE)
      facets.tcn.em.min = aggregate(r$facets.tcn.em,
                                    by = list(r$queryHits),
                                    FUN = min, na.rm=TRUE)
      facets.tcn.em.max = aggregate(r$facets.tcn.em,
                                    by = list(r$queryHits),
                                    FUN = max, na.rm=TRUE)
      
      # for each gene, populate total copy number: max, min and median of overlapping segments
      sample.df$facets.tcn.em.median[facets.tcn.em.median$Group.1] <- facets.tcn.em.median$x
      sample.df$facets.tcn.em.min[facets.tcn.em.min$Group.1] <- facets.tcn.em.min$x
      sample.df$facets.tcn.em.max[facets.tcn.em.max$Group.1] <- facets.tcn.em.max$x
      
      
      # minor copy number: median, min
      # mior copy number is missing for some segments (NA), and the following 
      facets.lcn.em.median = aggregate(r$facets.lcn.em,
                                       by = list(r$queryHits),
                                       FUN = median, na.rm=TRUE)
      facets.lcn.em.min = aggregate(r$facets.lcn.em,
                                    by = list(r$queryHits),
                                    FUN = min, na.rm=TRUE)
      
      # for each gene, populate minor copy number: max, min and median of overlapping segments
      sample.df$facets.lcn.em.median[facets.lcn.em.median$Group.1] <- facets.lcn.em.median$x
      sample.df$facets.lcn.em.min[facets.lcn.em.min$Group.1] <- facets.lcn.em.min$x
      
      # aggregate Pier's LOH assessment
      facets.LOH.pier = aggregate(r$isLohPier,
                                  by = list(r$queryHits),
                                  FUN = min, na.rm=TRUE)
      sample.df$facets.loh.pier[facets.LOH.pier$Group.1] <- facets.LOH.pier$x
      
    } 
    
    
    
    
  }
  # add to data frame
  call.list[[full_sample_name]] <- sample.df
  
}

Facets <- bind_rows(call.list)

Facets[is.infinite(Facets$facets.lcn.em.min),"facets.lcn.em.min"]<- NA


```

```{r, createSNiPDxObj,results='hide'}

snipdx <- createSNiPDxObj(Stats = Stats,
                          Variant = Variant,
                          Facets = Facets,
                          Inventory = Inventory,
                          OrigSegments= OrigSegments,
                          trial = params$currTrial)

# extract dfs out of slots
dfIntegInventory <- snipdx@dfIntegInventory

FacetsSelected <- snipdx@Facets
StatsSelected <- snipdx@Stats
VariantSelected = snipdx@Variant
Calls = snipdx@Calls
Segments <- snipdx@Segments
Seg_VariantSelected <- snipdx@Seg_VariantSelected
Seg_Calls <- snipdx@Seg_Calls
Seg_dfInventory <- snipdx@Seg_dfInventory

### ==== export ==== ###
if(params$currTrial != "Non-Clinical"){
  folder_name <- paste0(processingDir, "/results/", Sys.Date(), "_", params$currTrial,
                        "_", "SNiPDx_by_Patients")
}else{
  folder_name <- paste0(processingDir, "/results/", Sys.Date(), "_", params$currTrial,
                        "_", "SNiPDx_by_Projects")
}

script_name <- paste(Sys.Date(),params$currTrial, "SNiPDx", params$currPatient,
                     sep = "_")

allScriptName <- paste(Sys.Date(),params$currTrial, 
                                               "SNiPDx", "Complete",
                                               sep = "_")

dataPath <- paste0(processingDir, "/data/")
dir.create(paste0(processingDir, "/data/"), showWarnings = F, recursive = T)
dataPath <- paste0(dataPath, allScriptName, "_Inventory_Processed_Data.rds")

if(!file.exists(dataPath)){
  saveRDS(snipdx, file = dataPath)
}

### ==== create an Complete (ALL samples) folder if one does not exist ==== ###

alloutFolder <- paste0(folder_name, "/", allScriptName,  "_results/")

# create and store s4 object and inventory in /results/ and /data/
if(!dir.exists(alloutFolder)){
  dir.create(alloutFolder,showWarnings = F,recursive = T)
  
  inventoryList <- list("gene_level_inventory" = dfIntegInventory,
                        "segment_level_inventory" = Seg_dfInventory)
  
  write.xlsx(inventoryList,paste0(alloutFolder,allScriptName,"_Inventory.xlsx"),overwrite = T)
  
  saveRDS(snipdx, file = paste0(alloutFolder,allScriptName,"_Inventory_Processed_Data.rds"))
  
  ### run_info
  # specify newest sample ID and patient added to run_info
  new_batch <- tail(str_sort(unique(snipdx@dfIntegInventory$SEQ)),n = 1)
  
  if(params$currTrial != "Non-Clinical"){
    new_samples <- snipdx@dfIntegInventory %>%
      filter(SEQ %in% new_batch) %>%
      select(`Sample Sequencing Name`, SUBJID) %>%
      distinct() %>%
      arrange(SUBJID)
      
    run_info <- data.frame(Trial = params$currTrial,
                           Last.Batch = new_batch,
                           new_samples = new_samples$`Sample Sequencing Name`,
                           SUBJID = new_samples$SUBJID,
                           Inventory = params$inInventory,
                           inSamplesFolder = params$inSamplesFolder,
                           inVendorData = params$inVendorData, 
                           panalAnnotations = params$panalAnnotations,
                           inExtensions = params$inExtensions,
                           Date = snipdx@date)
  }else{
    new_samples <- snipdx@dfIntegInventory %>%
    filter(SEQ %in% new_batch) %>%
    select(`Sample Sequencing Name`, Project) %>%
    distinct() %>%
    arrange(Project)
    
  run_info <- data.frame(Trial = params$currTrial,
                         Last.Batch = new_batch,
                         new_samples$`Sample Sequencing Name`,
                         new_samples$Project,
                         Inventory = params$inInventory,
                         inSamplesFolder = params$inSamplesFolder,
                         inVendorData = params$inVendorData, 
                         panalAnnotations = params$panalAnnotations,
                         inExtensions = params$inExtensions,
                         Date = snipdx@date)
  }
  
  write.csv(run_info,file = paste0(alloutFolder,allScriptName,"_run_info.csv"))
}

```


```{r,subset_patient, results='hide'}
### ==== Subset individual patients by patient ID or Project if Nonclinical:  ====
if(params$currTrial == "Non-Clinical"){
  dfIntegInventory <- dfIntegInventory %>%
  filter(`Project` %in% params$currPatient)
  
  Seg_dfInventory <- Seg_dfInventory %>% 
    filter(`Project` %in% params$currPatient)
  
} else{
  # if currSampleType is not All types, subset the inventory df by the specified currSampleType
  dfIntegInventory <- dfIntegInventory %>%
    filter(SUBJID %in% params$currPatient)
  
  Seg_dfInventory <- Seg_dfInventory %>% 
    filter(SUBJID %in% params$currPatient)
}

```


```{r, splitSamples,results='hide'}

# if current trial is nonclinical, split by Project
if(params$currTrial == "Non-Clinical"){
  
  ID <- unique(sort(dfIntegInventory$`Project`))
  
  #relabel NA as string NA's
  dfIntegInventory$`Project` <- dfIntegInventory$`Project` %>%
    replace_na("NA")
  
   split <- split(dfIntegInventory, list(dfIntegInventory$`Project`,
                                        dfIntegInventory$`Sample Sequencing Name`),
                 drop = TRUE)
  
}else{
  ### if else rp3500-01, rp6306-01, rp6306-02
  
  # subset and group sample by patient id, sample type, and seq name into nested list
  ID <- unique(sort(dfIntegInventory$SUBJID))

  # split by patient level and sample types/ patient
  # not all patients will have both PBMC and Tumor samples
  split <- split(dfIntegInventory, list(dfIntegInventory$SUBJID,
                                        dfIntegInventory$SAMPTYPE2,
                                        dfIntegInventory$`Sample Sequencing Name`),
                 drop = TRUE)

}

# shorten subsetted df names to match headers during knit
shortenSplitName <- paste(sub("\\_GER.*|\\_GEC.*", "", names(split)), # extract everything before GER* or GEC* 
                          sub('.*\\_', '', names(split)), # extract everything after the last "_"
                          sep = "_" )

names(split) <- shortenSplitName

# store df subsets by patient ID
IDSplit <- lapply(ID, function(x){
  split[grepl(x, names(split))]
})

names(IDSplit) <- ID

```


```{r,include = FALSE}
  DT::datatable(matrix())
```

```{r, revised, results='asis' ,fig.width=20,fig.height=18,out.width="50%"}
### Gene lvl summary table
summarySelectedTables <- list(Facets = FacetsSelected, 
                      Stats = StatsSelected,
                      # CNV = CNVLong, 
                      Variants = VariantSelected,
                      Calls = Calls)

### segment lvl summary table
segmentSelectedTables <- list(Segments = Segments, 
                              `Segment Variants`= Seg_VariantSelected,
                              `Segment Calls` = Seg_Calls,
                              Complete = Seg_dfInventory)

for(patient in names(IDSplit)){
  # print patient headers
  cat(sprintf("\n\n# %s\n\n", patient))
  cat("\n\n")
  for(sampleType in names(IDSplit[[patient]])){
    
     # extract sample ID for each patient/sample type
      currSampleID <- 
        unique(IDSplit[[patient]][[sampleType]]$`Sample Sequencing Name`)
      
      # shorten the sequencing name, append to header/dfs 
    shortenSampleID <- paste(sub("\\_GER.*|\\_GEC.*", "", currSampleID), # extract everything before GER* or GEC* 
                          sub('.*\\_', '', currSampleID), # extract everything after the last "_"
                          sep = "_" )
    
    IDSplit[[patient]][[sampleType]]$`Short Sample Sequencing Name` <- 
      shortenSampleID
    
      
    cat(sprintf("\n\n## %s\n\n", paste(sampleType)))
      cat("\n\n")
      
      cat(paste("GCP path to bam file:", unique(IDSplit[[patient]][[sampleType]]$GCP_path_to_bam)))
      
       cat("\n\n")
      
      # print Summary Table headers per patient/sample pair
      cat(sprintf("\n\n### %s\n\n", "Gene Level Summary Tables {.tabset .tabset-pills}"))
      cat("\n\n")
      
      summaryTableList <- lapply(summarySelectedTables, function(currDF){
        #currDF[currDF$`Sample Sequencing Name` %in% currSampleID,]
        currDF %>%
          filter(`Sample Sequencing Name` %in% currSampleID) %>%
          mutate(`Short Sample Sequencing Name` = shortenSampleID, .before = 1) %>%
          distinct() %>%
          select(-`Sample Sequencing Name`) # remove long sample IDs, keep original in "Complete" DF
      })
      
      summaryTableList[["Complete"]] <- IDSplit[[patient]][[sampleType]]

      # # #create sub table headers
      for(table in names(summaryTableList)){
        cat(sprintf("\n\n#### %s\n\n", table))
        cat("\n\n")

         cat(knitr::knit_print(DT::datatable(summaryTableList[[table]],
                                     width = "100%",
                                     height = "auto",
                                 rownames = FALSE,
                                 options = list( scrollX = TRUE))))

         cat("\n\n")
         cat("\n\n")
         cat("\n\n")
         cat("\n\n")
      }

      # print plot headers per patient/sample pair
      cat(sprintf("\n\n### %s\n\n", "Plots {.tabset .tabset-pills}"))
      cat("\n\n")
      
      # list all plot paths per sample ID
      imagePaths <- list.files(
        path = paste0(params$inSamplesFolder, "/"),
                 pattern = glob2rx(paste0(currSampleID, "*.png")),
                 recursive = T,
                 include.dirs = T,
                 full.names = T)
      
      # pull from REPARE facets only
      imagePaths <- imagePaths[grep("*/FACETS/REPARE/facets/*",imagePaths)]
      
      # omit facets from purecn
      imagePaths <- imagePaths[grep("*facets_with_purecn_diplogr*",imagePaths, invert = T)]

      
      # select for non QC image Paths
      imagePaths <- imagePaths[grepl("\\.rawsummary|\\.facets", imagePaths)]
      
      #omit spider.facets
      imagePaths <- imagePaths[!grepl("spider.facets", imagePaths)]
      
      #remove png paths from ./reports folder (duplicates)
      imagePaths <- imagePaths[!grepl("/report/", imagePaths)]
         
      
      for(path in imagePaths){
        #print(path)
        ### label each plot's name extracted 
        plotName <- strsplit(x = path,split = "/")[[1]]
        plotName <- plotName[length(plotName)]
        plotName <- plotName %>%
          gsub(".png", "", .) %>%
          gsub(currSampleID, "", .) %>%
          gsub("^\\.", "\\1", .)
        #print(plotName)
        
        cat(sprintf("\n\n#### %s\n\n", plotName))
        cat("\n\n")
          
        #knitr::include_graphics(path) #only works if rmd is in same folder as png folder
        # cat(paste0("![](","../", path, ")"))
        cat(paste0("![](", path, ")"))
        
  
      }
      ### ==== segment lvl knits ==== ####
      #if there is segments data read in for curr Patient and Sample ID
      currSeg_dfInventory <- segmentSelectedTables$Complete %>%
        dplyr::filter(`Sample Sequencing Name` == currSampleID)
      
      if(nrow(currSeg_dfInventory) > 0){
        cat(sprintf("\n\n### %s\n\n", "Segment Level Summary Tables {.tabset .tabset-pills}"))
        cat("\n\n")
      
      # #create sub table headers
        for(table in names(segmentSelectedTables)){
          cat(sprintf("\n\n#### %s\n\n", table))
          cat("\n\n")


           segSummaryTableList <- lapply(segmentSelectedTables, function(currDF){
             currDF %>% filter(`Sample Sequencing Name` %in% currSampleID) %>%
            mutate(`Short Sample Sequencing Name` = shortenSampleID, .before = 1) %>%
            distinct() %>%
            relocate(`Sample Sequencing Name`,.after = last_col())
           })

           cat(knitr::knit_print(DT::datatable(segSummaryTableList[[table]],
                                       width = "100%",
                                       height = "auto",
                                   rownames = FALSE,
                                   options = list( scrollX = TRUE))))

           cat("\n\n")
           cat("\n\n")
           cat("\n\n")
           cat("\n\n")
        }
        
        # print plot headers per chromosome plot if seg contains variant, amplification or HomDel
        cat(sprintf("\n\n### %s\n\n", "Chromosome Plots {.tabset .tabset-pills}"))
        cat("\n\n")
        
        chrom_df  <- currSeg_dfInventory %>% 
          filter(!is.na(AAChange) | Seg_CNV %in% c("amp", "homdel")) %>% 
        distinct(chrom) 
        
        # add chr 19 to all chrom plots
        # if there are no chrom pos, only append chr19
        if(nrow(chrom_df) >0){
          chr <- c(paste0("chr", chrom_df$chrom), "chr19")
          chr <- unique(chr)
        }else{
          chr <-paste0("chr", "19")
        }
      
        #order 
        chr <- sort(as.integer(str_remove(chr, "chr")))
        chr <- paste0("chr", chr)

        
         # list all plot paths per sample ID
        # chrPaths <- list.files(path =  paste0(params$inSamplesFolder, 
        #                                      currSampleID, "/facets/chromosome_plots/"),
        #            pattern = paste(chr, collapse = "|"),
        #            recursive = T,
        #            include.dirs = T,
        #            full.names = T)
        
         chrPaths <- list.files(path =  paste0(params$inSamplesFolder),
                   #pattern = paste(paste(currSampleID, chr,sep = "." ), collapse = "|"),
                  pattern =  paste0(currSampleID, ".", chr, "\\.", collapse = "|"),
                   recursive = T,
                   include.dirs = T,
                   full.names = T)
         
         # pull from REPARE facets only
        chrPaths <- chrPaths[grep("*/FACETS/REPARE/facets\\b",chrPaths)]
        
        # order
        chrPaths <- chrPaths[order(nchar(chrPaths))]
        
        
        for(currChr in seq_along(chrPaths)){
          #print(path)
          ### label each plot's name extracted 
          
          cat(sprintf("\n\n#### %s\n\n", chr[currChr]))
          cat("\n\n")
          
          cat(paste0("![](", chrPaths[currChr], ")"))
          
    
      }
      } # end segment
  }
}
```


```{r, save,eval=TRUE}
##### project specific inventory tables
outFolder <- paste0(folder_name, "/", script_name, "_results/")
dir.create(outFolder,showWarnings = F,recursive = T)

inventoryList <- list("gene_level_inventory" = dfIntegInventory,
                      "segment_level_inventory" = Seg_dfInventory)
 
write.xlsx(inventoryList,paste0(outFolder,script_name,"_Inventory.xlsx"),overwrite = T)

```