# arguments
# pileup file
# PON file
# sample QC (Good, Fair or Poor)

#!/usr/bin/env Rscript
library("optparse")
library(facets)
library(reshape2)


base_dir <- normalizePath(getwd(), mustWork = TRUE)


option_list = list(
    make_option(c("-p", "--pileup_fn"), type="character", help="pileup file (generated by snp-pileup)", metavar="character"),

    make_option(c("-r", "--reference_PON"), type="character",
              help="reference panel of normals .RData file", metavar="character"),

    make_option(c("-g", "--panel_genes_file"), type="character",
              help="reference panel of normals .RData file", metavar="character"),

    make_option(c("-i", "--sample_id"), type="character",
              help="sample ID", metavar="character"),

    make_option(c("-o", "--sample_output_folder"), type="character",
              help="sample ID", metavar="character"),

    make_option(c("-d", "--sample_insert_distribution"), type="character",
              help="sample ID", metavar="character"),

    make_option(c("-c", "--cluster_info"), type="character",
              help="sample ID", metavar="character"),

    make_option(c("-m", "--min_cov"), type="numeric", default='500',
              help="sample ID", metavar="numeric"),

    make_option(c("-h", "--het_threshold"), type="numeric", default='0.05',
              help="threshold for allele fraction to distingish heterozygous SNPs", metavar="numeric"),

    make_option(c("-s", "--secondary_normalization"), type="logical", default=TRUE,
              help="whether to normalize the logR singal (after normal adjustment), by distance to the primer, using lowess regression", metavar="numeric"),

    make_option(c("-l", "--cval"), type="numeric", default=50,
              help="facets cval for segmentation", metavar="numeric"),

    make_option(c("-k", "--diplogr_file"), type="character", default=NULL,
              help="file with diplogr value from purecn", metavar="character"),

    make_option(c("-a", "--deltaCN"), type="numeric", default=0.2,
              help="minimum detectable difference in CN from diploid state", metavar="numeric"),

    make_option(c("-y", "--sample_type"), type="character", default="FFPE",
              help="blood or FFPE. Whether the DNA sample is from blood or FFPE tissue", metavar="numeric"),

    make_option(c("-b", "--snp.nbhd"), type="character", default=250,
              help="Facets paramer. Controls sampleing of SNPs. Distance between adjacent SNPs", metavar="numeric"),

    make_option(c("-e", "--only_hets"), type="logical", default=FALSE,
              help="Whether to use hets only", metavar="numeric"),

    make_option(c("-u", "--utils_path"), type="character", default="src/utils",
              help="Path to the utility scripts", metavar="numeric"),

    make_option(c("-n", "--normalize_is"), type="logical", default=TRUE,
              help="normalize by local insert sizes", metavar="numeric"),

    make_option(c("-z", "--sample_is_fn"), type="character",
              help="file with local insert sizes", metavar="numeric")

)

opt_parser = OptionParser(option_list=option_list, add_help_option=FALSE);
opt  = parse_args(opt_parser);

stopifnot(!is.null(opt$pileup_fn))
stopifnot(!is.null(opt$sample_output_folder))

source(file.path(base_dir, opt$utils_path, 'plot.facets.R'))
source(file.path(base_dir, opt$utils_path, 'drawRawSample.R'))
source(file.path(base_dir, opt$utils_path, 'drawRawSampleGenes.R'))
source(file.path(base_dir, opt$utils_path, 'plotIS.R'))
source(file.path(base_dir, opt$utils_path, 'classifySample.R'))
source(file.path(base_dir, opt$utils_path, 'getSamplePrimerDistance.R'))
source(file.path(base_dir, opt$utils_path, 'normalizeWrtPrimer.R'))

source(file.path(base_dir, opt$utils_path, 'plotSampleModified.R'))
source(file.path(base_dir, opt$utils_path, 'findClosestSamples.R'))
source(file.path(base_dir, opt$utils_path, 'plotCoverageScatterplot.R'))

source(file.path(base_dir, opt$utils_path, 'logRlogORspider.mod.R'))
source(file.path(base_dir, opt$utils_path, 'writeBw.R'))
source(file.path(base_dir, opt$utils_path, 'plotGC.R'))
source(file.path(base_dir, opt$utils_path, 'extract.ei.bias.R'))

source(file.path(base_dir, opt$utils_path,  'normalize_by_is.R'))
source(file.path(base_dir, opt$utils_path,  'plotISanalysis.R'))

# load the panel of normals reference
# in particular, data frame pon.df
load(opt$reference_PON)

# prepare output folders
plotFolderMain = opt$sample_output_folder
dir.create(plotFolderMain, showWarnings = FALSE)

# decide on cluster of the sample
load(opt$cluster_info) # loads km, cluser.df
r2<- plotIS(opt$sample_insert_distribution, cluster.centers, cluser.ffpe.and.blood.df, draw=TRUE, plot_fn=paste0(plotFolderMain, gsub('_D1.*', '',opt$sample_id), '.inserts.distribution.png'))
sample.class <- classifySample(r2$v, cluster.centers, cluser.ffpe.and.blood.df, opt$sample_type, plot=NULL)
print(paste('Sample Insert Distribution class:', sample.class))


######
# load the pileup
facets.pileup <- opt$pileup_fn
print(facets.pileup)
zz=gzfile(facets.pileup,'rt')
pl=read.csv(zz,header=TRUE, sep=',', stringsAsFactors = FALSE)
close(zz)

# this pileup will save as normal, and therefore needs 'File2' prefix
colnames(pl) <- gsub('File1', 'File2', colnames(pl))
sampleloci <- paste(pl$Chromosome, pl$Position)
#decide with reference set to use, depending on quality of sample in question
class.used <- cluser.ffpe.and.blood.df[sample.class, 'use.pon.cluster']
includeCol <- paste0('include.cluster.', class.used)


# for each SNP, mark whether it will be used (based on sample QC-dependent column)
sampleloci <- paste(pl$Chromosome, pl$Position)
pl$logR <- log((pl$File2A + pl$File2R+1)/pon.df[as.character(sampleloci),'coverageMean'])
pl$is.target <- NA
pl$is.target <- pon.df[as.character(sampleloci),'is.target']
pon.df$target.type[grepl('rs',pon.df$target)] <- 'SNP'
pon.df$target.type[grepl('exon',pon.df$target)] <- 'exon'
pl$target.type <- NA
pl$target.type <- pon.df[sampleloci, 'target.type']
pl$snipdx.gene <- pon.df[sampleloci, 'snipdx.gene']
pl$primerMinDistance <- pon.df[sampleloci, 'primerMinDistance']
pl$closestPrimerIsSnpTiled <- pon.df[sampleloci, 'closestPrimerIsSnpTiled']
sample.max.primer.distance <- getSamplePrimerDistance(pl, minCov=opt$min_cov, plot_fn=paste0(plotFolderMain, opt$sample_id, '.coverage.vs.primerDistance.png'))

pl$af <- pl$File2A/(pl$File2A + pl$File2R) # allele fractuibs
pl$is.het <- (pl$af>opt$het_threshold) & (pl$af<(1-opt$het_threshold))
pl$total.cov <- pl$File2R + pl$File2A

pl$include <- pon.df[sampleloci, includeCol]
pl$is.variable <- FALSE
pl$is.variable <- pon.df[sampleloci, 'is.variable']
pl$include <- pl$include &
        ((!is.na(pl$is.variable)) & (!pl$is.variable)) &
        (pl$Chromosome %in% paste0('chr',1:22)) &
        (pl$is.target | ((pl$is.het) & (!is.na(pl$primerMinDistance)) & (pl$primerMinDistance<sample.max.primer.distance)) )

# in this mode, only use heterozygous SNPs for analysis
if (opt$only_hets) {
    pl$include <- pl$include & pl$is.het
}

# draw raw sample data
drawRawSample(pl, opt$sample_id, plotFolderMain, smoothWind=21)

# only keep restricted SNPs
pl <- subset(pl, include==TRUE)

# count approximate number of heterozygotes
nHets <- sum(pl$is.het, na.rm=TRUE)
print(paste('hets:', nHets))
# IDs of only included SNPs
sampleloci.inc <- paste(pl$Chromosome, pl$Position)

# create a new version of the pileup, normalized using the PONs
# call it pl2
pl2 <- pl[,c('Chromosome', 'Position', 'Ref', 'Alt')]
pl2$Chromosome = substr(pl2$Chromosome, 4,100 )
# use counts_r and counts_a
# simulating a normal sample
# the fake normal sample
# this is coming from the PONs
class.used <- cluser.ffpe.and.blood.df[sample.class, 'use.pon.cluster']
pon.column <- paste0('cov.cluster.',class.used)
pl2$File1R = floor(pon.df[sampleloci.inc , pon.column]/2)
pl2$File1A = floor(pon.df[sampleloci.inc , pon.column]/2)
pl2$File1E = 0
pl2$File1D = 0
#concatenate columns from new and old pileups
# File 1 is normal sample (here faked using normal samples)
# File 2 is tumor sample
# R is reads reporting reference allele
# A is reads reporting alternative allele
pl2 <- cbind(pl2, pl[,c('File2R','File2A','File2E','File2D', 'primerMinDistance', 'closestPrimerIsSnpTiled')])
pl2$r <- ((pl2$File2R + pl2$File2A)/ median(pl2$File2R + pl2$File2A, na.rm=TRUE)) / ((pl2$File1R + pl2$File1A) / median(pl2$File1R + pl2$File1A, na.rm=TRUE))
pl2 <- subset(pl2, !is.na(r) & !is.infinite(r))

# establish distance from primer, where for this sample, the coverage drops below a threshold
r<- getSamplePrimerDistance(pl2, minCov=100, plot_fn=paste0(plotFolderMain, opt$sample_id, '.logr.vs.primerDistance.png'), specifyCol='r', flagCol='closestPrimerIsSnpTiled')

# normalize by local insert sizes
if (opt$normalize_is) {
    if (file.exists(opt$sample_is_fn)) {
        sample.is <- read.csv(opt$sample_is_fn)
        rownames(sample.is) <- paste(sample.is$Chromosome, sample.is$Position)
        is_plot_fn<- paste0(plotFolderMain, opt$sample_id, '.is.comp.png')
        pl2.columns <- colnames(pl2)
        pl2 <- normalize_by_is(pl2, sample.is, pon.df, is_plot_fn)
        pl2 <- pl2[,pl2.columns]
    }
}

# normalziation of the LogR signal, by distance from the primer
if (opt$secondary_normalization) {
    pl2.norm <- normalizeWrtPrimer(pl2, plot_fn=paste0(plotFolderMain, opt$sample_id, '.logr.primer.normalization.png'))
    pl2 <- pl2.norm[,colnames(pl2)]
}

mean.cov <- round(mean(pl2$File2R + pl2$File2A, na.rm=TRUE))
plotCoverageScatterplot(pl2, plot_fn=paste0(plotFolderMain, opt$sample_id, '.cov.scatterplot.png'))

# save the transformed pileup file
fn <- paste0( opt$sample_output_folder, opt$sample_id, '.pileup.with.pon.csv')
write.csv(pl2[,1:(ncol(pl2)-3)], file=fn, row.names=FALSE, quote=FALSE)

# now run facets, on such a transformed file
# Facets execution
###################################################
rcmat <- readSnpMatrix(fn)

xx <- preProcSample(rcmat, unmatched=TRUE, het.thresh=opt$het_threshold, ndepthmax=40000, deltaCN=opt$deltaCN, snp.nbhd=opt$snp.nbhd)

# read in file with diplogr from purecn
if (opt$diplogr_file != "NA") {
    diplogr <- read.delim(file = opt$diplogr_file, header = FALSE)
    diplogr_value <- diplogr[1, 1]
    oo <- procSample(xx, cval = opt$cval, dipLogR = diplogr_value)
} else {
    oo <- procSample(xx, cval = opt$cval)
}

fit=emcncf(oo)
# end of Facets
###################################################

# standard Facets plot
png(paste0(plotFolderMain, opt$sample_id, '.facets.png'), width=2000, height=1300, res = 200)
options(repr.plot.width=15, repr.plot.height=10)
plotSampleModified(x=oo,emfit=fit,
                   sname = paste(substr(opt$sample_id,1,30 ), 'purity=', round(fit$purity,2),'ploidy=', round(fit$ploidy,2), 'insert cluster:', sample.class, 'cov:',mean.cov,'X, max primer-SNP distance',sample.max.primer.distance,'bp hets: ', nHets ))
dev.off()

fit$cncf$loc.start <- fit$cncf$start
fit$cncf$loc.end <- fit$cncf$end
fit$cncf$ploidy <- fit$ploidy
fit$cncf$purity <- fit$purity
fit$cncf$chr <- fit$cncf$chrom
fit$cncf$loc.start <- fit$cncf$start
fit$cncf$loc.end <- fit$cncf$end
fit$cncf$major.cn <- fit$cncf$tcn.em - fit$cncf$lcn.em
fit$cncf$major.cn[is.na(fit$cncf$lcn.em )]<- fit$cncf$tcn.em[is.na(fit$cncf$lcn.em )]
fit$cncf$loh <- fit$cncf$lcn.em==0

# fill NAs in the LOH column (Facets occasionally leaves them when there are too few SNPs in the segment)
median.logr = median(fit$cncf$mafR)
std.logr = sd(fit$cncf$mafR)
# this is Pier's formula. If mafR is an outlier, assume LOH in the segment
if (sum(is.na(fit$cncf$lcn.em))>0) {
    fit$cncf$loh[is.na(fit$cncf$lcn.em)] <- FALSE
    fit$cncf$loh[is.na(fit$cncf$lcn.em) & (fit$cncf$mafR> (median.logr + std.logr)) ] <- TRUE
}
# the facets table
dominik_txt_fn = paste0(plotFolderMain, opt$sample_id, '.cnv.table.txt')
write.table(fit$cncf, file=dominik_txt_fn, sep='\t', quote = FALSE, row.names=FALSE)

# write Facets intermediate files
save(rcmat, oo, fit, file=paste0(plotFolderMain, opt$sample_id, '.facets.RData'))

tumor.median <- median(xx$pmat$rCountT, na.rm=TRUE)
normal.median <- median(xx$pmat$rCountN, na.rm=TRUE)

# write a bigwig file for troubleshooting
bw.fn <- paste0(plotFolderMain, opt$sample_id, '.targets.norm.logR.wig')
writeBw(bw.fn, opt, xx)
jointseg.fn <- paste0(plotFolderMain, opt$sample_id, '.jointseg.csv')
write.csv(xx$jointseg, file=jointseg.fn)

# plot Facets style
plot.facets(dominik_txt_fn, paste0(opt$sample_id, '.ascatlike'), plotFolderMain, format='png', onlyAutosomal=TRUE,
            sample.label=paste(opt$sample_id, 'Ploidy=',round(fit$ploidy,2), 'Purity=',round(fit$purity,2), 'insert cluster:', sample.class, ' hets: ', nHets ), ylim3=c(0,5), majorCol='major.cn')

# load and order panel genes, by chromosome coordinate
panel.genes <- read.table(opt$panel_genes_file, header=TRUE, stringsAsFactors=FALSE)
panel.genes.oreded <- panel.genes[order(as.integer(panel.genes$chromosome_name), panel.genes$transcript_start),]

# make visualizations with various amount of magnification
# draw all-chromosome summaries
drawRawSampleGenes(pl, opt$sample_id, plotFolderMain, panel.genes.oreded, fit$cncf, oo, margin=5e6, allChroms=TRUE)
drawRawSampleGenes(pl, opt$sample_id, plotFolderMain, panel.genes.oreded, fit$cncf, oo, margin=1e5)
drawRawSampleGenes(pl, opt$sample_id, plotFolderMain, panel.genes.oreded, fit$cncf, oo, margin=2e6)
# draw a gene per page summary
drawRawSampleGenes(pl, opt$sample_id, plotFolderMain, panel.genes.oreded, fit$cncf, oo, margin=1e5, allChroms=FALSE)

# draw Facet's spider plot
png(paste0(plotFolderMain, opt$sample_id, '.spider.facets.png'), width=2000, height=1000, res = 200)
par(mfrow=c(1,2))
logRlogORspider.mod(oo$out, oo$dipLogR)
logRlogORspider.mod(oo$out, oo$dipLogR, xlim=c(-0.5,0.5), ylim=c(0,0.6))
dev.off()

# Plot GC bias
gc_fn <- paste0(plotFolderMain, opt$sample_id, '.gc.png')
gc_csv_fn <- paste0(plotFolderMain, opt$sample_id, '.gc.csv')
plotGC(gc_fn, gc_csv_fn, xx)

# plot exons vs SNP bias
ei.bias.plot.fn <- paste0(plotFolderMain, opt$sample_id, '.bias.png')
ei.bias.table.fn <- paste0(plotFolderMain, opt$sample_id, '.bias.csv')
extract.ei.bias(ei.bias.plot.fn,ei.bias.table.fn, xx, opt)

# local insert size analysis
is.plot.fn <- paste0(plotFolderMain, opt$sample_id, '.is.outliers.png')
if (!is.null(opt$sample_is_fn) && file.exists(opt$sample_is_fn)) {
  plotISanalysis(xx, pon.df, opt$sample_is_fn, is.plot.fn)

}


